File: C#/src/readtrack/Readtrack.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

/// <summary>
/// readtrack.cs - Reads a LEGO Stunt Rally track file. By Yellowberry, MIT license.
/// </summary>

namespace srtools
{
    class Readtrack
    {
        public static string lego_header;
        public static int crashint, trk_size, trk_theme, trk_time, trk_fsize, iters;
        public static long realsize;

        public static byte[][][][] trk_pieces;

        static void Main(string[] args)
        {
            string trk_name = "";
            try
            {
                trk_name = args[0];
            }
            catch (IndexOutOfRangeException)
            {
                Console.Beep();
                Console.WriteLine("Hey, you gotta provide an argument for the track!");
                System.Environment.Exit(1);
            }
            FileStream file = File.Open(trk_name, FileMode.Open, FileAccess.Read);
            using (BinaryReader reader = new BinaryReader(file))
            {
                realsize = file.Length;
                lego_header = bstr(reader.ReadBytes(12));
                crashint = bint(reader.ReadBytes(4));
                trk_fsize = bint(reader.ReadBytes(4));
                if (trk_fsize != 65576) throw new Exception("Binary filesize reports non 64KB file!");
                else if (trk_fsize != realsize) throw new Exception("Incorrect filesize, corrupted track!");
                trk_size = bint(reader.ReadBytes(4));
                trk_theme = bint(reader.ReadBytes(4));
                trk_time = bint(reader.ReadBytes(4));

                iters = 8 * (trk_size + 1);
                int skip = (56 - (trk_size * 8)) * 16;
                trk_pieces = new byte[iters][][][];

                for (int y = 0; y < iters; y++)
                {
                    byte[][][] tmp = new byte[iters][][];
                    for (int x = 0; x < iters; x++)
                    {
                        byte[][] piece = new byte[3][];
                        reader.ReadBytes(4);
                        for (int z = 0; z < 3; z++)
                        {
                            byte[] bytes = reader.ReadBytes(4);
                            piece[z] = bytes;
                        }
                        tmp[x] = piece;
                    }
                    reader.ReadBytes(skip);
                    trk_pieces[y] = tmp;
                }

                Console.WriteLine("## Track information for {0} ##", trk_name);
                Console.WriteLine("Fizesize: {0}", trk_fsize);
                Console.WriteLine("Track Type: {0}", LSRutil.t_size[trk_size]);
                Console.WriteLine("Track Theme: {0}", LSRutil.t_theme[trk_theme]);
                Console.WriteLine("Time of Day: {0}", LSRutil.t_time[trk_time]);
                Console.WriteLine("---");
                PrettyPiece(0,0);
            }
        }

        static string bstr(byte[] data)
        {
            string str;
            if (!BitConverter.IsLittleEndian) Array.Reverse(data);
            str = Encoding.UTF8.GetString(data);
            return str;
        }

        static int bint(byte[] data)
        {
            int i;
            if (!BitConverter.IsLittleEndian) Array.Reverse(data);
            i = BitConverter.ToInt32(data, 0);
            return i;
        }

        static int bchar(byte data)
        {
            int ch;
            ch = data;
            return ch;
        }

        static float bfloat(byte[] data)
        {
            float fl;
            if (!BitConverter.IsLittleEndian) Array.Reverse(data);
            fl = BitConverter.ToSingle(data, 0);
            return fl;
        }

        public static byte[] Combine(params byte[][] arrays)
        {
            byte[] ret = new byte[arrays.Sum(x => x.Length)];
            int offset = 0;
            foreach (byte[] data in arrays)
            {
                Buffer.BlockCopy(data, 0, ret, offset, data.Length);
                offset += data.Length;
            }
            return ret;
        }

        public static int GetTheme(byte theme)
        {
            int t = -1;

            if      (theme == 0x3B) t = 0; // Jungle
	        else if (theme == 0x3F) t = 1; // Ice
            else if (theme == 0x47) t = 2; // Desert
            else if (theme == 0x43) t = 3; // City
	        return t;
        }

        public static int GetNid(int id, int theme)
        {
            int ret = 0;

            int[] tb = { };

            if      (theme == 0) tb = LSRutil.pi_jung; // Jungle
	        else if (theme == 1) tb = LSRutil.pi_ice;  // Ice
	        else if (theme == 2) tb = LSRutil.pi_dsrt; // Desert
	        else if (theme == 3) tb = LSRutil.pi_city; // City

            foreach (int i in tb)
            {
                if (i == id) break;
                ret += 1;
            }

            return ret;
        }

        public static int[] ParsePiece(byte[][] data)
        {
            float height = bfloat(data[0]);
	        if      (height == -1) height = 0;
	        else if (height ==  8) height = 1;
	        else if (height == 16) height = 2;
	        else if (height == 24) height = 3;

            byte[] piece = data[1];
            int theme = GetTheme(piece[1]);
            int id = piece[0];
            int nid = GetNid(id, theme);
            int rotation = piece[2];

            return new int[5] { nid, id, theme, rotation, (int)height };
        }

        public static void PrettyPiece(int x, int y)
        {
            int[] info = ParsePiece(trk_pieces[y][x]);
            Console.WriteLine("### Piece information for x{0},y{1} ###", x, y);
            Console.WriteLine("nID: {0}", info[0]);
            Console.WriteLine("ID: {0:X}", info[1]);
            Console.WriteLine("Theme: {0}", LSRutil.t_theme[info[2]]);
            Console.WriteLine("Rotation: {0}", info[3]);
            Console.WriteLine("Height: {0}", info[4]);
        }
    }
}

File: Python/readtrack.py
## readtrack.py - Reads a LEGO Stunt Rally track file. By Yellowberry, MIT license.

import os, sys, struct, array, random

import lsrutil as s # other python file

### Functions

def getTheme(theme):
	t = -1
	if   theme == b'\x3B': t = 0 # Jungle
	elif theme == b'\x3F': t = 1 # Ice
	elif theme == b'\x47': t = 2 # Desert
	elif theme == b'\x43': t = 3 # City
	return t

def getNid(id, theme):
	ret = 0
	if   theme == 0: tb = s.pi_jung # Jungle
	elif theme == 1: tb = s.pi_ice  # Ice
	elif theme == 2: tb = s.pi_dsrt # Desert
	elif theme == 3: tb = s.pi_city # City
	for h in tb:
		if h == id:
			break
		ret += 1
	return ret
 
def parsePiece(data):
	height = struct.unpack("<f", data[0])[0] # Unpack LE float
	if   height == -1: height = 0
	elif height ==  8: height = 1
	elif height == 16: height = 2
	elif height == 24: height = 3

	piece = data[1]
	theme = getTheme(piece[1:2])
	id = hex(piece[:1][0])
	nid = getNid(s.ib(int(id, 16)), theme)

	rotation = struct.unpack("<i", data[2])[0] # Unpack LE int

	#print(height, theme, id, nid)
	return [nid, id, theme, rotation, height]

def prettyPiece(x,y):
	info = parsePiece(trk_pieces[y][x])
	print("### Piece information for", "x" + str(x) + ",y" + str(y), "###")
	print("nID:", info[0])
	print("ID:", info[1])
	print("Theme:", s.t_theme[info[2]])
	print("Rotation:", info[3])
	print("Height:", info[4])

### Main logic

## Checking arguments
get_x = 0
get_y = 0
try:
	get_x = int(sys.argv[2])
	get_y = int(sys.argv[3])
except:
	pass # we take what we can get

## Opening file and collecting data

trk_name = os.path.basename(os.path.normpath(sys.argv[1]))
with open(sys.argv[1], "rb") as file:

	realsize = os.path.getsize(sys.argv[1]) # get actual filesize

	lego_header = file.read(12) # 'LEGO MOTO\0\0\0'
	crashint = struct.unpack("<i", file.read(4))[0] # Always 5
	trk_fsize = struct.unpack("<i", file.read(4))[0]
	if trk_fsize != 65576: raise Exception("Binary filesize reports non 64KB file!")
	elif realsize != trk_fsize: raise Exception("Incorrect filesize, corrupted track!")
	trk_size = struct.unpack("<i", file.read(4))[0] # 0 - 1
	trk_theme = struct.unpack("<i", file.read(4))[0] # 0 - 3
	trk_time = struct.unpack("<i", file.read(4))[0] # 0 - 1

	iters = 8*(trk_size+1)
	skip = (56-(trk_size*8))*16
	trk_pieces = []
	for x in range(iters):
		tmp = []
		for y in range(iters):
			piece = []
			file.read(4) # discard null bytes
			for z in range(3):
				piece.append(file.read(4))
			tmp.append(piece)
		file.read(skip)
		trk_pieces.append(tmp)	

## Printing info

#print(iters, skip)
#print(trk_pieces)
print("## Track information for", trk_name, "##")
print("Fizesize:", trk_fsize)
print("Track Type:", s.t_size[trk_size])
print("Track Theme:", s.t_theme[trk_theme])
print("Time of Day:", s.t_time[trk_time])
print("---")
prettyPiece(get_x,get_y)